### DISCUSSION TITLE
WASI module for IPC

### DISCUSSION THREAD

**Discussion by skyne98** [wasmerio/wasmer#2631] [Q&A]
Hello there!

### Summary

Is it possible to, in theory, create a separate WASI module for supporting IPC in both user and kernel lands? As for right now, WASI module system seems to be pretty undocumented. However, `kernel-net` seems like an extension of the standard, is it?

Thanks!

**Comment by ghost** [comment]
I am also interested in this.

Currently I use hashicorp plugin approach. It works but it's too verbose to use

**Comment by Hywan** [comment]
Hello @skyne98,

Can you elaborate a little bit about `kernel-net` please? I'm not aware of that.

**Comment by skyne98** [comment]
Hey @Hywan,

It's been quite a while since I have posted that message, but I still remember some bits. Here I was talking about `wasmer`'s backend, which was released back then. It was created to run WASM binaries inside the kernel space by a kernel module WASM runner. Running WASM in the kernel meant theoretically much faster system calls.

If I remember right, `kernel-net` was a Wasmer module/extension which allowed making network requests inside that runner (from the kernel land). I was hoping if it was possible to get a module for IPC communication, potentially through the kernel barrier to be able to write a WASM kernel module which will communicate with user-space WASM binaries (or other WASM kernel modules) and make ultra-fast system calls in the process.

**Comment by Hywan** [comment]
This work has been extracted into its own repository, https://github.com/wasmerio/kernel-wasm/.

It's an interesting question… which I can't answer for the moment. Let's keep this issue open and let's see if more people are interested by this, we could maybe make it a reality. Or if anyone wants to contribute… feel free :-).
